pkg/payload/precondition/clusterversion/upgradeable.go

```diff
@@ -82,6 +84,18 @@ func (pf *Upgradeable) Run(ctx context.Context, releaseContext precondition.Rele
 // Name returns Name for the precondition.
 func (pf *Upgradeable) Name() string { return "ClusterVersionUpgradeable" }
 
+// getCurrentVersion determines and returns the cluster's current version by iterating through the
+// provided update history until it finds the first version with update State of Completed.
+func getCurrentVersion(history []configv1.UpdateHistory) string {
+	for _, h := range history {
+		if h.State == configv1.CompletedUpdate {
+			klog.V(5).Infof("Cluster current version=%s", h.Version)
+			return h.Version
+		}
+	}
+	return ""
+}
```

Comment:
What about partial upgrades? I guess it changes the current version.

Reply:
We should look at how `oc get clusterversion` shows the current version.

Reply:
Any upgrade, even one that immediately fails preconditions gets inserted into the version history at index 0 and then that is used to check if upgrade is for a z-level upgrade. And as I said it is then comparing desired to itself. If I understand your question correctly, no, a partial update would not change my calculated "current version" because it would not have State Completed. I don't see how a cluster can be considered at a given version if the version upgrade was not completed. So it really comes down to what defines a cluster being at a given version. It can't simply be that the version shows up in the version history and I would hope we would not go down the path of trying to figure out how partial a partial upgrade was and if it got far enough along to consider the cluster at that version.

Reply:
> We should look at how `oc get clusterversion` shows the current version.

Gives the completed version and shows upgrade progress:

NAME      VERSION      AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.5.0-rc.1   True        True          6m17s   Working towards 4.5.0-rc.2: 70% complete

And when I stepped on this upgrade with another before it was completed (there are three entries in version history now):

NAME      VERSION      AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.5.0-rc.1   True        True          8m27s   Working towards quay.io/openshift-release-dev/ocp-release-nightly@sha256:cec724b008c35aaf36aeea8fdfc3f22f3f735d772d05f392acd86610de1608cd: downloading update

Reply:
@jottofar  I asked the question because when a cluster is partially updated, customers can try updating again to a newer version e.g. (in 4.4. fast channel)
* first upgrade attempted from 4.4.3 to 4.4.6. 
* next upgrade 4.4.6 to 4.4.10 which ended as a partial upgrade 
* The upgrade again attempted to 4.4.11. Now when CVO checks if there is an edge exists to 4.4.11 what current version it uses?

Reply:
Or the question which version it should use as current version? You are suggesting that we use 4.4.6 as the current version as the last successful update was to that version.

Reply:
I am not sure what qualifies for a partial upgrade? I am wondering if the operators like api server, etcd has upgraded (control plane) has moved to the next version the does it make any difference to us.

Reply:
> Or the question which version it should use as current version? You are suggesting that we should use 4.4.6 as the current version as the last successful upgrade was to that version.

In your example, yes. It's just like my example above it and that's what 'oc get currentversion' would show as well.

Reply:
To use a partial version we would have to introduce logic to figure out what makes a partial version "good enough" to be considered  the current version. And as I said, this would differ from what the system returns as current version now which is the last version with state Completed.